### ISSUE TITLE
parser errors are not included in TFLint's main output

### ISSUE THREAD

**Issue by carlosvsilva** [terraform-linters/tflint#983]
I'm trying to use it with Notepad++ linter plugin but I need checkstyle output. Jenkins could benefit from this too. But currently checkstyle output format seems completely broken. the only checkstyle output is constantly:

tflint /f checkstyle main.tf
```xml
<?xml version="1.0" encoding="UTF-8"?>
<checkstyle></checkstyle>Failed to load configurations. 3 error(s) occurred:

Error: Invalid character

  on main.tf line 28, in provider "aws":
  28:   qwery;

The ";" character is not valid. Use newlines to separate arguments and blocks, and commas to separate items in collection values.
...
```
```hcl
// Write code here
terraform {
  required_providers {
    docker = {
      source = "terraform-providers/docker"
    }
  }
}

provider "docker" {
  host = "unix:///var/run/docker.sock"
}

resource "docker_image" "nginx" {
  name         = "nginx:latest"
  keep_locally = false
}

resource "docker_container" "nginx" {
        image = docker_image.nginx.latest
        name  = "tutorial"
        ports {
                internal = 80
                external = 8000
        }
}
provider "aws" {
        qwery;
        sdlkjhfgldsajgf;

}

```
I also create a config C:\Users\username\\.tflint.hcl but it always fails to load some configuration.

Running from cmd.exe the binary straight from github, or in powershell from choco install

## Version
TFLint version 0.20.3 (powershell)
TFLint version 0.21.0 (cmd.exe)
Terraform v0.13.5

**Comment by bendrucker** [comment]
What is going on here?

```tf
provider "aws" {
        qwery;
        sdlkjhfgldsajgf;
}
```

TFLint fails to load your config, starting with `;` in that provider block. You can argue whether it should print anything to stdout if the config is invalid but I don't think this is specific to the checkstyle formatter, might appear with json as well. As long as the exit code is non-zero it seems to be working as intended.

**Comment by carlosvsilva** [comment]
Thats an error on purpose otherwise how do you test an error detector aka linter ?

**Comment by bendrucker** [comment]
TFLint is not intended as a syntax validator. It is meant to run on valid Terraform configuration. If your configuration is not parse-able, TFLint errors early. While it is possible to convert those errors into TFLint errors for output, that is not the current behavior, regardless of formatter. Right now the error is just returned as a string.

**Comment by carlosvsilva** [comment]
But checkstyle format is something like this:

<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProblems.js"><error line="1" column="10" severity="error" message="&apos;addOne&apos; is defined but never used. (no-unused-vars)" source="eslint.rules.no-unused-vars" /><error line="2" column="9" severity="error" message="Use the isNaN function to compare with NaN. (use-isnan)" source="eslint.rules.use-isnan" /><error line="3" column="16" severity="error" message="Unexpected space before unary operator &apos;++&apos;. (space-unary-ops)" source="eslint.rules.space-unary-ops" /><error line="3" column="20" severity="warning" message="Missing semicolon. (semi)" source="eslint.rules.semi" /><error line="4" column="12" severity="warning" message="Unnecessary &apos;else&apos; after &apos;return&apos;. (no-else-return)" source="eslint.rules.no-else-return" /><error line="5" column="1" severity="warning" message="Expected indentation of 8 spaces but found 6. (indent)" source="eslint.rules.indent" /><error line="5" column="7" severity="error" message="Function &apos;addOne&apos; expected a return value. (consistent-return)" source="eslint.rules.consistent-return" /><error line="5" column="13" severity="warning" message="Missing semicolon. (semi)" source="eslint.rules.semi" /><error line="7" column="2" severity="error" message="Unnecessary semicolon. (no-extra-semi)" source="eslint.rules.no-extra-semi" /></file></checkstyle>

Shouldn't tflint output this kind of filled in xml when run with switch -f checkstyle ?

**Comment by bendrucker** [comment]
See my explanation above. Yes that is an example of checkstyle output. When TFLint encounters invalid configuration that it cannot load, it prints the error as text and exits early. Doesn't matter what the cause is, TFlint only formats the results of its rules. This could be changed but is currently working as designed and occurs regardless of the chosen formatter.

**Comment by wata727** [comment]
When I first implemented this, I remember wondering how to output unintended errors in the checkstyle format. As he says, I think we should distinguish between issues for rules and runtime errors. It seems that the `<error />` in the checkstyle format is intended to represent an issue. If the checkstyle format can represent runtime errors separately from the issues, it can be followed, but I cannot find such the format.

Instead, TFLint prints always valid checkstyle-formatted XML to stdout. If there is stderr, I think it may be possible to process the result correctly by stopping the parsing of stdout. What do you think?

BTW, the JSON format can output issues and runtime errors separately, and can always output a valid JSON unless it is a fatal error. This is because we are free to define the JSON format :)

**Comment by bendrucker** [comment]
I think if the errors have diagnostics attached (filenames and line numbers) it's reasonable to try to convert them for output. For an error resulting from anything but loading the config, continuing to print to stderr and exit 1 seems correct. But perhaps not printing to stdout at all.

**Comment by carlosvsilva** [comment]
If I understand you correctly, input file errors should be printed as checkstyle XML to stdout, while internal tflint errors should be printed to stderr ? That seems the logical approach to me :)

**Comment by wata727** [comment]
Yes. It should already behave like that.

**Comment by carlosvsilva** [comment]
I agree it should. Can you release that functionality in the next version? :)

**Comment by bendrucker** [comment]
We're officially going in circles now. We've thoroughly described the currently behavior and rationale and also alluded to how it might be made to work differently in cases such as yours. If you'd like to step up and submit a PR that'd be awesome. If you want to wait until we get around to it that's ok too, but we certainly can't make promises about when that might be.

**Comment by carlosvsilva** [comment]
Ok I just feel that checkstyle is not currently supported as far as I can tell. I don't program, I work in terraform ops.

Besides isn't it trivial to just do in tflint:
print '<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="' + fullFilename +'"><error line="' + currentLineNo + '" column="' + currentColumn + '" severity="' + msgType + '" message="' + escapeToHTML(messageString) + '" source="' + ruleName + '" />' ?

**Comment by bendrucker** [comment]
I've locked this thread to avoid further distraction. Thanks for reporting the issue, there's a genuine usability improvement to be had here.

The pressure tactics are not cool. If you are unwilling to contribute a pull request, sit tight. Asking for priority, trivializing the work, and tossing up pseuo-code is quite rude, regardless of your intentions.
